### ISSUE TITLE
[Snackbar] Issue with the handling of the action button when dealing with multiple snackbars.

### ISSUE THREAD

**Issue by CyborgSemon** [material-components/material-components-web#6540]
Ive got a question / suggestion about the snackbar.
By the way that the API appears to work, it seems as if every time you want a new message in the snackbar, it needs to be its own snackbar component. This is because of how snackbar closing is handled with events and the action button. If I have a snackbar with an action, I have to call `.listen` on the snackbar variable if I want the action button to be able to do anything. This seems fine at first.

Lets say I want my snackbar action button to call an `undo` function when clicked. I would have to:
(Im using the CDN version of MDC in these examples)
```html
<div id="snackbar" class="mdc-snackbar">
  <div class="mdc-snackbar__surface">
    <div class="mdc-snackbar__label">Item moved to favorites</div>
    <div class="mdc-snackbar__actions">
      <button type="button" class="mdc-button mdc-snackbar__action">
        <div class="mdc-button__ripple"></div>
        <span class="mdc-button__label">Undo</span>
      </button>
    </div>
  </div>
</div>
```
```js
const snackbar = new mdc.snackbar.MDCSnackbar(document.querySelector("#snackbar"));

snackbar.listen("MDCSnackbar:closing", (eve)=> {
  if (eve.detail.reason == "action") {
    // Undo function
  }
});

// snackbar.open();
```

This seems great and all if I only have one snackbar. However there is a high likely hood of having multiple snackbars with different messages and different actions in an application.
So lets say I have 3 different snackbar message and action button combos that I need to show my user. My initial thought of how to use the `.listen()` feature would just to make 3 separate HTML snackbars and just listen to them individually. This however would get very messy very quickly as the application got bigger.

Luckily the API does let you change the message and the action button text of a snackbar using `.labelText = "..."` and `.actionButtonText = "..."` properties  respectively, but there doesnt seem to be a way to change the actual action of the action button. So how am I supposed to differentiate between the different action calls using a single snackbar component?

The logical option would be to make a wrapper API around a single snackbar component in the HTML. So we would end up with something like this

```html
<div id="snackbar" class="mdc-snackbar">
  <div class="mdc-snackbar__surface">
    <div class="mdc-snackbar__label"></div>
    <div class="mdc-snackbar__actions">
      <button type="button" class="mdc-button mdc-snackbar__action">
        <div class="mdc-button__ripple"></div>
        <span class="mdc-button__label"></span>
      </button>
    </div>
  </div>
</div>
```
```js
const snackbar = new mdc.snackbar.MDCSnackbar(document.querySelector("#snackbar"));
let currentActionFunc = ()=> {};

// Actions:
// - undoFavorites
// - retry
// - undoDelete

snackbar.listen("MDCSnackbar:closing", (eve)=> {
  if (eve.detail.reason == "action") {
    currentActionFunc();
  }
});

function snack(msg = "", actionText = "", actionFunc = ()=> {}) {
  currentActionFunc = actionFunc;

  snackbar.labelText = msg;
  snackbar.actionButtonText = actionText;

  snackbar.open();
}

// snack("...", "Undo", ()=> {/* Custom function */});
// snack("...", "Retry", ()=> {/* Custom function */});
// snack("...", "Undo", ()=> {/* Custom function */});
```
Which to me seems a little annoying because you now have to keep track of another global variable.

Is there a proper way to do this? Or have I pretty much nailed the head on the coffin as to the expected behaviour for handling this. Because I fell that there should be a property in the API that allows you to set a function to the action button that just gets called automatically when the action button is clicked.

You could of course mutate the snackbar variable and just embed the action function into it `snackbar.actionClick = ()=> {};`. However doing that feels a little "hacky" as it doesn't really feel right to "edit" the API. But the API documentation doesn't mention any proper / official way to handle multiple actions soooo... here we are :P 

---

TLDR: Is this this expected way to handle multiple snackbars? Or can I be doing this better?
And if this issue hasn't been thought of, what changes to the API should be made to make this flow more seamless.

(Also while I'm at at, I have the exact same issue with the dialog component. But I'll probably get into that in another Github issue as there are some other use cases that are a little different and come with their own "hacky" solutions)

**Comment by allan-chen** [comment]
Hi Simon,

I believe the component was built such that each distinct snackbar has its own template & listener (essentially self-contained), and dialog should follow this pattern as well. 

For now, would recommend following the one template per snackbar approach pattern, or the global variable workaround if there's really significant bloat with the intended approach.

I can discuss the possibility of a feature request to bind some function to the action with the team later this week (i.e. something like `snackbar.actionHandler = ()=>{}`), so I'll keep this open in the meantime.

**Comment by CyborgSemon** [comment]
Sweet that would be great thanks.

Yeah that makes sense. You can definitely see it becoming an issue if you have like, more than 5 snackbars just sitting at the bottom of the index page. Add that with any number of dialog options as well and it can get very bloated.

```html
<div id="dialog1" class="mdc-dialog">...</div>
<div id="dialog2" class="mdc-dialog">...</div>
<div id="dialog3" class="mdc-dialog">...</div>
<div id="dialog4" class="mdc-dialog">...</div>
<div id="dialog5" class="mdc-dialog">...</div>

<div id="snackbar1" class"mdc-snackbar">...</div>
<div id="snackbar2" class"mdc-snackbar">...</div>
<div id="snackbar3" class"mdc-snackbar">...</div>
<div id="snackbar4" class"mdc-snackbar">...</div>
<div id="snackbar5" class"mdc-snackbar">...</div>
<div id="snackbar6" class"mdc-snackbar">...</div>
```

Do you want me to make another issue outlining my dialog concerns as well? Or just add it onto this issue as it is essentially the same concept but with a few additions.

**Comment by allan-chen** [comment]
Curious about the use case for that many dialogs/snackbars in a single page, but I agree that with vanilla it could become a little populated & this is where a framework can come in to help with conditional rendering.

Nevertheless we can just use this issue & I'll respond after discussion with team.

**Comment by CyborgSemon** [comment]
Yeah this is an SPA with quite a few conformations and options to undo your last action. But its also just handy to have a wrapper API that I can just copy past in any other project that needs snackbars.

The same sort of case is with the dialog. Except the dialog is a little less adaptable as there are no properties to quickly change the title, content or action buttons. (We will ignore the full screen variant as thats currently not available for web)
So my 2 issues with the dialog are:

1. No easy way to change title & content (Though this can be done by the dev without too much effort. See code example below)

Suggestion is to have a `.title` and .`content` property that can be accessed kind of like how the snackbar does it. And have the `.content` one also have a ability to accept HTML text or JS DOM elements as there are cases where you might want radio buttons in the dialog for example.


2. The action buttons cant easily be changed. Though they are an interesting case because according to the design spec there can be anywhere between 0 and 3 action buttons in each dialog which proves a bit of a challenge as if there are no action buttons, how is the dialog supposed to know when the user clicks a piece of content inside the dialog. (The [simple dialog ](https://material-components.github.io/material-components-web-catalog/#/component/dialog)for example)

Suggestion is to have a `.actions` property hold an array and the array holds either the buttons or the information that makes up the buttons. So it could look something like this:
```js
const dialog = new mdc.dialog.MDCDialog(document.querySelector("#dialog"));

dialog.actions = [
  {
    text: "Cancel",
    handler: ()=>{},
    default: true // Would act like the "data-mdc-dialog-button-default" data variable
  },
  {
    text: "Discard",
    handler: ()=> {
      // Do whatever you need to in here
    }
];
```
This way you can have as many or as little buttons as you want. How the handler would work is it would get called then the dialog close function would get called immediately after.

In the case of this array being empty, I'm assuming that the dialog would handle its closing the same way that the simple dialog handles its closing. Since that example doesnt have any action buttons.

---

Unfortunately my suggestions will most likely involve a rather large refactor of the dialog component, so I'm not really expecting any of them to go past the discussion phase. But it does kind of show off the "static-ness" of the component that feels like its designed in a way so that you dont have to include the same boilerplate code a bunch of times to get the functionality you want.


My "hacky" way to get the functionality described if your interested.
```html
<div id="dialog" class="mdc-dialog">
  <div class="mdc-dialog__container">
    <div class="mdc-dialog__surface">
      <h2 class="mdc-dialog__title" id="diagTitle"></h2>
      <div class="mdc-dialog__content" id="diagContent"></div>
      <div class="mdc-dialog__actions" id="diagButtons"></div>
    </div>
  </div>
  <div class="mdc-dialog__scrim"></div>
</div>
```

```js
// createElement is just a function I have that returns an HTML Element

function diag(title = "", content = createElement({type: "div"}), buttonsInfo = []) {
  document.querySelector("#diagTitle").innerHTML = title;
  
  dialog.content_.innerHTML = "";
  dialog.content_.appendChild(content);
  
  let buttonsDiv = document.querySelector("#diagButtons");
  buttonsDiv.innerHTML = "";
  
  let buttons = mdcDiagButtons(...buttonsInfo);
  for (let button of buttons) buttonsDiv.appendChild(button);
  
  mdc.autoInit(dialog.root);
  dialog.open();
}

function mdcDiagButtons() {
  let returnArr = [];
  for (let buttonInfo of arguments) {
    returnArr.push(createElement({type: "button", classes: ["mdc-button", "mdc-dialog__button"], children: [
      createElement({type: "div", classes: ["mdc-button__ripple"]}),
      createElement({type: "span", classes: ["mdc-button__label"], innerHTML: buttonInfo.text})
    ], onclick: buttonInfo.click, "data-mdc-auto-init": "MDCRipple"}));
  }
  
  return returnArr;
}

/*
diag(`Delete "Thing"?`, createElement({type: "div", innerHTML: "Are you sure you want to delete <b>Thing</b>?"}), [
  {
    text: "Cancel",
    click: ()=> {
      dialog.close();
    }
  },
  {
    text: "Delete",
    click: ()=> {
      // Delete thing here
      dialog.close();
    }
  }
]);
*/
```

So I pretty much just end up remaking the whole buttons section from scratch.

---

I can definitely see why the dialog is built the way it is as it is definitely a more complex use case than the snackbar. But I still dont think its nessacery to include `<div class="mdc-dialog__scrim"></div>` `x` amount of times in the same file for example.

Thanks for reading. Let me know how the discussion with the team goes :)

**Comment by allan-chen** [comment]
After discussing with the team, we're leaning towards intended behavior for this one. The dialog/snackbar instances were purposed to be immutable (which arguably could be less bug-prone), and it is generally not recommended to repurpose them. If the concern is lots of DOM nodes, we recommend writing a wrapper that instantiates/destroys snackbars or dialogs as needed; the templating capabilities of a web framework might also be useful here, though it may be overkill depending on the scale of the application.

**Comment by CyborgSemon** [comment]
Ok fair enough. Thanks for the update :)
